lib/veri.ml

```diff
@@ -0,0 +1,207 @@
+open Core_kernel.Std
+open Bap.Std
+open Bap_traces.Std
+module Bap_value = Value
+open Bil.Result
+open Monad.State
+open Trace
+
+let do_nothing () = return ()
+
+type point = {
+  code : Chunk.t;
+  side : event list
+}
+let is_first_read ctxt ev =
+  match Bap_value.get Event.register_read ev with
+  | None -> false
+  | Some mv -> Option.is_none (ctxt#lookup (Move.cell mv))
+
+let is_first_load ctxt ev = 
+  match Bap_value.get Event.memory_load ev with
+  | None -> false
+  | Some mv ->
+    match Seq.find ~f:(fun (v,_) -> v = ctxt#mem) ctxt#bindings with
+    | None -> true
+    | Some (_,s) ->
+      match value s with
+      | Bil.Mem storage -> Option.is_none (storage#load (Move.cell mv))
+      | _ -> true 
+
+let is_init_event ctxt ev =
+  is_first_read ctxt ev || is_first_load ctxt ev
+
+class context report arch = object(self:'s)
+  inherit Veri_traci.context arch
+  val report : Veri_report.t = report
+  val events : event list = []
+  val selves = None
```

Comment:
why `None`? It is always known, first of all. Second `None` is singular, `selves` is plural, didn't you find it wrong?